src/frontend/src/optimizer/plan_node/logical_join.rs

```diff
@@ -1099,6 +1103,135 @@ impl LogicalJoin {
         }
     }
 
+    #[allow(dead_code)]
+    fn to_stream_temporal_join(
+        &self,
+        predicate: EqJoinPredicate,
+        ctx: &mut ToStreamContext,
+    ) -> Result<PlanRef> {
+        assert!(predicate.has_eq());
+
+        let left = self.left().to_stream_with_dist_required(
+            &RequiredDist::shard_by_key(self.left().schema().len(), &predicate.left_eq_indexes()),
+            ctx,
+        )?;
+
+        if !left.append_only() {
+            return Err(RwError::from(ErrorCode::NotSupported(
+                "Temporal join needs a append-only left input".into(),
+                "Please ensure your left input is append-only".into(),
+            )));
+        }
+
+        let right = self.right();
+        let Some(logical_scan) = right.as_logical_scan() else {
+            return Err(RwError::from(ErrorCode::NotSupported(
+                "Temporal join needs a table scan as its lookup table".into(),
+                "Please provide a table scan".into(),
+            )));
+        };
```

Comment:
So, maybe we must choose one? 
1. need to introduce a special logical node, or the optimizations can not ensure the plan node here(suppose if we have an agg-join rule).
2. need to support any kind of batch query as the right input in the temporal join executor implementation.

Reply:
IMO, the temporal join is similar to the lookup join which can only look up an existing table and this is the main use case. Supporting other operators as the temporal join right input seems unrealistic.